MoreLinq.Test/TransposeTest.cs

```diff
@@ -0,0 +1,232 @@
+                                       .Where(x => IsPrime(x))
+                                       .Take(3)
+                                       .Select((x, i) => i == 1 ? MoreEnumerable.Generate(x, n => n * x).Take(2)
+                                                                : MoreEnumerable.Generate(x, n => n * x));
+
+            var result = matrix.Transpose().Take(5);
+
+            var expectations = new[]
+            {
+                new[] { 2,    3,    5 },
+                new[] { 4,    9,   25 },
+                new[] { 8,        125 },
+                new[] { 16,       625 },
+                new[] { 32,      3125 }
+            };
+
+            AssertMatrix(result, expectations);
+        }
+
+        [Test]
+        public void TransposeOrderSequencesAreIteratedIsIrrelevant()
+        {
+            var matrix = new[]
+            {
+                new[] { 10, 11 },
+                new[] { 20 },
+                new int[0],
+                new[] { 30, 31, 32 }
+            };
+
+            var transpose = matrix.Transpose().ToList();
+
+            transpose[1].AssertSequenceEqual(11, 31);
+            transpose[0].AssertSequenceEqual(10, 20, 30);
+            transpose[2].AssertSequenceEqual(32);
+        }
+
+        [Test]
+        public void TransposeInnerSequencesAreConsumedLazies()
+        {
+            var matrix = new[]
+            {
+                MoreEnumerable.From(() => 10, () => 11),
+                MoreEnumerable.From(() => 20, () => 22),
+                MoreEnumerable.From(() => 30, () => throw new InvalidOperationException(), () => 31),
+            };
+
+            var result = matrix.Transpose();
+
+            result.ElementAt(0).AssertSequenceEqual(10, 20, 30);
+
+            Assert.Throws<InvalidOperationException>(() =>
+                result.ElementAt(1));
+        }
+
+        [Ignore("Test about 100% laziness, the current approach do not support it")]
+        public void TransposeSequencesAreLazies()
+        {
+            var matrix = new[]
+            {
+                new[] { 10, 11, 12 },
+                new[] { 30, 31, 32 }.Select<int, int>(x => throw new Exception())
+            };
+
+            var first = matrix.First();
+            var count = first.Count();
+
+            matrix.Transpose().Take(count).ForEach((seq, i) =>
+            {
+                Assert.That(seq.First(), Is.EqualTo(first.ElementAt(i)));
+            });
+        }
+
+        static bool IsPrime(int number)
+        {
+            if (number == 1) return false;
+            if (number == 2) return true;
+
+            var boundary = (int)Math.Floor(Math.Sqrt(number));
+
+            for (var i = 2; i <= boundary; ++i)
+            {
+                if (number % i == 0)  return false;
+            }
+
+            return true;
+        }
+
+        public static void AssertMatrix<T>(IEnumerable<IEnumerable<T>> result, IEnumerable<IEnumerable<T>> expectation)
```

Comment:
This can be private. Also, assertion methods usually take expectations first then actual results so would be best to swap the arguments for sake of consistency.